README.md
[![Build Status](https://travis-ci.org/bazaarvoice/emodb.svg?branch=master)](https://travis-ci.org/bazaarvoice/emodb)

EmoDB
=====

Store your feelings here.

Written by [Bazaarvoice](http://www.bazaarvoice.com): see [the credits page](https://github.com/bazaarvoice/emodb/blob/master/Credits.md) for more details.

Introduction
------------

EmoDB is a RESTful HTTP server for storing JSON objects and for watching for changes
to those events.

It is designed to span multiple data centers, using eventual consistency (AP) and multi-master
conflict resolution.  It relies on [Apache Cassandra] (http://cassandra.apache.org/) for
persistence and cross-data center replication.

Documentation
-------------

[Release Notes] (markdowns/ReleaseNotes.md)

[System of Record API] (markdowns/SystemOfRecordApi.md)

[System of Record Deltas] (markdowns/Deltas.md)

[Databus API] (markdowns/DatabusApi.md)

[BlobStore API] (markdowns/BlobStoreApi.md)

[Queue API] (markdowns/QueueServiceApi.md)

[Stash API] (markdowns/StashApi.md)

[Hadoop and Hive Support] (markdowns/MapReduce.md)

[Operations] (markdowns/Operations.md)

[EmoDB SDK] (markdowns/Maven.md)

[API Keys] (markdowns/Security.md)

[Legacy Java Client Support] (markdowns/LegacyJavaClients.md)

Quick Start
-----------

### Installation

#### Using binaries:

1. Download the tarball from [EmoDB binaries] (https://github.com/bazaarvoice/emodb/releases)

2. Run the Emodb server locally. This will start zookeeper and cassandra locally.
    
        $ ./start-local.sh
        ...
        INFO  [2012-05-14 19:12:19,802] org.eclipse.jetty.server.AbstractConnector: Started InstrumentedBlockingChannelConnector@0.0.0.0:8080
        INFO  [2012-05-14 19:12:19,805] org.eclipse.jetty.server.AbstractConnector: Started SocketConnector@0.0.0.0:8081
        # Use Ctrl-C to kill the server when you are done.

3.  Check that the server responds to requests (from another window):

        $ curl -s "http://localhost:8081/ping"
        pong

        $ curl -s "http://localhost:8081/healthcheck"
        * deadlocks: OK
        * emo-cassandra: OK
          127.0.0.1(127.0.0.1):9160 879us

4.  To erase the EmoDB data, simply delete the data folder:

        $ rm -rf bin/data/
        $ bin/start-local.sh

#### Using source code:

1. Download the [EmoDB source code] (https://github.com/bazaarvoice/emodb):

        $ git clone git@github.com:bazaarvoice/emodb.git emodb

2.  Build the source and run the tests:

        $ cd emodb
        $ mvn clean install

3.  Run the EmoDB server locally:

        $   cd web-local
        $ ./start.sh
        ...
        INFO  [2012-05-14 19:12:19,802] org.eclipse.jetty.server.AbstractConnector: Started InstrumentedBlockingChannelConnector@0.0.0.0:8080
        INFO  [2012-05-14 19:12:19,805] org.eclipse.jetty.server.AbstractConnector: Started SocketConnector@0.0.0.0:8081
        # Use Ctrl-C to kill the server when you are done.

4.  Check that the server responds to requests (from another window):

        $ curl -s "http://localhost:8081/ping"
        pong

        $ curl -s "http://localhost:8081/healthcheck"
        * deadlocks: OK
        * emo-cassandra: OK
          127.0.0.1(127.0.0.1):9160 879us

5.  To erase the EmoDB data and restart with a clean slate:

        $ cd web-local
        $ ./start-clean.sh

### Quick Tutorial

The following examples assume you have [jq] (https://stedolan.github.io/jq/) or an equivalent (see Recommended
Software below).  It is optional-- `jq .` just formats the JSON responses to make them easier to read.

Note that EmoDB comes with a built-in API Key Management system. For local testing, and tutorials, we have enabled "anonymous" access, 
which allows us to skip sending API Keys. This is done to make examples more readable. 
See [Security Management] (https://github.com/bazaarvoice/emodb/blob/master/markdown/SecurityManagement.md) for details.  

1.  Create a table in the System of Record.  Specify a "table template" with properties that will be returned with
    every object in the table:

        $ curl -s -XPUT -H "Content-Type: application/json" \
                "http://localhost:8080/sor/1/_table/review:testcustomer?options=placement:'ugc_global:ugc'&audit=comment:'initial+provisioning',host:aws-tools-02" \
                --data-binary '{"type":"review","client":"TestCustomer"}' | jq .
        {
          "success": true
        }

2.  Verify that the table was created as expected.  The result should be the table template.

        $ curl -s "http://localhost:8080/sor/1/_table/review:testcustomer" | jq .
        {
          "client": "TestCustomer",
          "type": "review"
        }

3.  Via the Databus, subscribe to changes on all tables containing reviews:

        $ curl -s -XPUT -H "Content-Type: application/x.json-condition" \
            "http://localhost:8080/bus/1/demo-app" \
            --data-binary '{..,"type":"review"}' | jq .
        {
          "success": true
        }

4.  Store a document in the System of Record:

        $ curl -s -XPUT -H "Content-Type: application/json" \
            "http://localhost:8080/sor/1/review:testcustomer/demo1?audit=comment:'initial+submission',host:aws-submit-09" \
            --data-binary '{"author":"Bob","title":"Best Ever!","rating":5}' | jq .
        {
          "success": true
        }

5.  Update the document in the System of Record:

        $ curl -s -H "Content-Type: application/x.json-delta" \
            "http://localhost:8080/sor/1/review:testcustomer/demo1?audit=comment:'moderation+complete',host:aws-cms-01" \
            --data-binary '{..,"status":"APPROVED"}' | jq .
        {
          "success": true
        }

6.  See what the document looks like after the update:

        $ curl -s "http://localhost:8080/sor/1/review:testcustomer/demo1" | jq .
        {
          "~deleted": false,
          "~firstUpdateAt": "2012-06-22T20:11:53.473Z",
          "~id": "demo1",
          "~lastUpdateAt": "2012-06-22T20:12:09.679Z",
          "~signature": "7db2ef78f7830acaaa53f242a5e5ffa1",
          "~table": "review:testcustomer",
          "~version": 2,
          "author": "Bob",
          "client": "TestCustomer",
          "rating": 5,
          "status": "APPROVED",
          "title": "Best Ever!",
          "type": "review"
        }

7.  Look at the first 10 documents in the table, sorted arbitrarily:

        $ curl -s "http://localhost:8080/sor/1/review:testcustomer" | jq .
        [
          {
            "~deleted": false,
            "~firstUpdateAt": "2012-06-22T20:11:53.473Z",
            "~id": "demo1",
            "~lastUpdateAt": "2012-06-22T20:12:09.679Z",
            "~signature": "7db2ef78f7830acaaa53f242a5e5ffa1",
            "~table": "review:testcustomer",
            "~version": 2,
            "author": "Bob",
            "client": "TestCustomer",
            "rating": 5,
            "status": "APPROVED",
            "title": "Best Ever!",
            "type": "review"
          }
        ]

8.  Poll the Databus to see the pending change events.  Note that generally there will be one event per update, but multiple
    updates to the same entity may be consolidated.  The current complete object is returned with each event.  The result will
    look something like:

        $ curl -s "http://localhost:8080/bus/1/demo-app/poll?ttl=30" | jq .
        [
          {
            "eventKey":"e9f5b640-caf8-11e1-96fe-0013e8cdbb13#review:testcustomer#demo1",
            "content":{
              "~deleted":false,"~lastUpdateAt":"2012-07-11T01:37:02.372Z","~id":"demo1",
              "~lastUpdateAt":"2012-07-11T01:37:13.351Z","~table":"review:testcustomer",
              "~signature": "7db2ef78f7830acaaa53f242a5e5ffa1","~version":2,
              "author":"Bob","client":"TestCustomer","rating":5,
              "status":"APPROVED", "title":"Best Ever!","type":"review"
            }
          }
        ]

9.  Acknowledge one of the Databus events to indicate we don't need it any more (copy the ID from the previous response):

        $ curl -s -XPOST -H "Content-Type: application/json" \
            "http://localhost:8080/bus/1/demo-app/ack" \
            --data-binary '["e9f5b640-caf8-11e1-96fe-0013e8cdbb13#review:testcustomer#demo1"]' | jq .
        {
          "success": true
        }

10.  Look at the timeline showing all System of Record changes to the object (modulo compaction):

        $ curl -s "http://localhost:8080/sor/1/review:testcustomer/demo1/timeline?audit=true" | jq .
        [
          {
            "timestamp": "2012-07-11T01:37:13.351+0000",
            "id": "f080f970-caf8-11e1-96fe-0013e8cdbb13",
            "delta": "{..,\"status\":\"APPROVED\"}",
            "audit": {
              "comment": "moderation complete",
              "host": "aws-cms-01",
              "~sha1": "4507332be7b42bd100a233be3847e5df99fbeb2d"
            }
          },
          {
            "timestamp": "2012-07-11T01:37:02.372+0000",
            "id": "e9f5b640-caf8-11e1-96fe-0013e8cdbb13",
            "delta": "{\"author\":\"Bob\",\"rating\":5,\"title\":\"Best Ever!\"}",
            "audit": {
              "comment": "initial submission",
              "host": "aws-submit-09",
              "~sha1": "33aef50cae4e44cc7be803054335bafdd375644b"
            }
          }
        ]

Recommended Software
--------------------

For debugging, it's useful to have a JSON pretty printer.  On a Mac with [Homebrew] (http://mxcl.github.com/homebrew/)
installed:

    brew install jq
    
Alternatively, you can use jsonpp
    
    brew install jsonpp

Alternatively, use Python's `json.tool`:

    alias jsonpp='python -mjson.tool'

Many of the examples include `jq` or `jsonpp`.  Running the examples without `jsonpp` will work just fine, but the results may
be more difficult to read.


markdowns/ReleaseNotes.md
Release Notes
=============

## 0.76

Backward-incompatible change to the `DataStore.getTimeline()` REST API and Java API.  The previous version, designed
for debugging use only, was inherently unscalable and could cause EmoDB to fail with OutOfMemory errors when objects
in the System of Record had too many very large uncompacted deltas.  The new version resolves this by using a streaming
approach that never loads all deltas into memory at once, although working with objects with many uncompacted deltas
will be very slow and is *strongly* discouraged.

The `DataStore.getTimeline()` method now takes optional `start`, `end`, `reversed` and `limit` parameters to allow
retrieving subsets of record history and/or paging through history using multiple requests.  The default `limit` is 10.

## 0.73

Added a new utility class `com.bazaarvoice.emodb.sor.api.Coordinate` that clients may use to simplify their code.

## 0.72

Extended `DataStoreStreaming` to wrap the DataStore `updateAll` method to reduce the amount of retry work done when
the connection to the server is lost unexpectedly so as to support essentially unlimited #s of records streamed in a
single call.

## 0.71

Added the following HTTP client library helper classes that wrap the DataStore and BlobStore streaming read methods
`listTables`, `scan`, `getSplit` and `scanMetadata` to automatically detect and recover from interruptions in the
streaming due to unexpected server connection loss so as to support essentially unlimited #s of records streamed in
a single call.

    com.bazaarvoice.emodb.sor.client.DataStoreStreaming
    com.bazaarvoice.emodb.blob.client.BlobStoreStreaming

## 0.70

Upgraded EmoDB client and server to Dropwizard 0.6.1, Jackson 2.1.x and the open source Ostrich 1.5.2.

The REST API is unchanged.  You may continue to use pre-0.70 clients with 0.70+ servers and vice-versa. However,
upgrading client code to the 0.70+ client libraries will likely require upgrading your clients to Dropwizard 0.6.1
and the open source Ostrich.  This will involve code changes since Dropwizard and Ostrich both have significant API
changes.

## 0.69

The System of Record `Conditions.in()` clause will now perform in O(1) instead of O(n). In combination with the new
method `Conditions.intrinsic(String, Condition)`, it should be reasonably efficient to subscribe to a large set of
enumerated SoR tables. Databus subscriptions that previously used `or(intrinsic(~table:..),intrinsic(~table:..),..)`
(like Polloi) will be converted automatically to the more efficient O(1) `intrisic(~table:..,..,..,..)` form when
possible--clients should not need to make any changes.

## 0.67

If using the EmoDB Java HTTP client libraries, the `createTable()` and `dropTable()` REST APIs on `DataStore`
and `BlobStore` may now be called from any AWS region, not just `us-east-1`.

Previously, these specific APIs would fail with an exception in any AWS region other than `us-east-1`. Now the
servers return an HTTP 301 redirect instead, redirecting the client to the `us-east-1` ELB. The EmoDB Java clients
will follow the redirect automatically but, because `createTable()` and `dropTable()` use the HTTP `PUT` and `DELETE`
methods, most other HTTP clients will not. Those clients will need Emo-specific software updates.

## 0.66

Added `Date BlobMetadata.getTimestamp()` to the Blob Store API. It returns the date/time when a blob was uploaded.


yum/start-local.sh
#!/bin/bash

#
# Starts the following servers locally:
# - EmoDB (ports 8080, 8081)
# - Cassandra (port 9160)
# - ZooKeeper (port 2181)
#
# The first time this is run, Cassandra will be initialized with a default
# schema and an empty # data set.  Data will be stored in "bin/data".
# On subsequent runs where "bin/data/" already exists, the Cassandra
# schema and data will not be modified.
#
# Once the server is running you can access Emodb using the following commands:
# Note: For local testing, we have enabled anonymous access. In production you should be using API Keys. Read more about it in "API Keys" section.
#
#   curl localhost:8081/ping
#   curl localhost:8080/sor/1/_table
#
# For production, use the web jar not web-local, and also you shouldn't be providing cassandra.yaml or -z switch since you wouldn't want to start a local cassandra or zookeeper
# java -jar bin/emodb-web.jar server conf/config-local.yaml conf/config-ddl-local.yaml
#

java -jar bin/emodb-web-local-*.jar server conf/config-local.yaml conf/config-ddl-local.yaml conf/cassandra.yaml -z
